homeassistant/components/bosch_alarm/services.py

```diff
@@ -0,0 +1,76 @@
+"""Services for the bosch_alarm integration."""
+
+from __future__ import annotations
+
+import asyncio
+import datetime as dt
+from typing import Any
+
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntryState
+from homeassistant.core import HomeAssistant, ServiceCall
+from homeassistant.exceptions import HomeAssistantError, ServiceValidationError
+from homeassistant.helpers import config_validation as cv
+from homeassistant.util import dt as dt_util
+
+from .const import ATTR_CONFIG_ENTRY_ID, ATTR_DATETIME, DOMAIN, SERVICE_SET_DATE_TIME
+from .types import BoschAlarmConfigEntry
+
+
+def validate_datetime(value: Any) -> dt.datetime:
+    """Validate that a provided datetime is supported on a bosch alarm panel."""
+    date_val = cv.datetime(value)
+    if date_val.year < 2010:
+        raise vol.RangeInvalid("datetime must be after 2009")
+
+    if date_val.year > 2037:
+        raise vol.RangeInvalid("datetime must be before 2038")
+
+    return date_val
+
+
+SET_DATE_TIME_SCHEMA = vol.Schema(
+    {
+        vol.Required(ATTR_CONFIG_ENTRY_ID): cv.string,
+        vol.Optional(ATTR_DATETIME): validate_datetime,
+    }
+)
+
+
+def setup_services(hass: HomeAssistant) -> None:
+    """Set up the services for the bosch alarm integration."""
+
+    async def async_set_panel_date(call: ServiceCall) -> None:
```

Comment:
Note: you can now use `call.hass` property to access `hass`, so services can be moved to top-level module functions and no longer need to be nested inside the registration function

Reply:
https://github.com/home-assistant/core/pull/145306